doc/radosgw/vault.rst

```diff
@@ -0,0 +1,79 @@
+===========================
+HashiCorp Vault Integration
+===========================
+
+HashiCorp `Vault`_ can be used as a secure key management service for
+`Server-Side Encryption`_ (SSE-KMS).
+
+#. `Vault authentication`_
+#. `Create a key in Vault`_
+#. `Configure the Ceph Object Gateway`_
+#. `Upload object`_
+
+Vault authentication
+====================
+
+Vault provides several authentication mechanisms. Currently, the Object Gateway
+supports the `token authentication method`_ only.
+
+When authenticating with Vault using the token method, save the token in a
+plain-text file. The path to this file must be provided in the Gateway
+configuration file (see below). For security reasons, ensure the file is
+readable by the Object Gateway only.
+
+Create a key in Vault
+=====================
+
+Generate and save a 256-bit key in Vault. Vault provides several Secret
+Engines, which store, generate, and encrypt data. For instance, create a key
+in the `KV Secrets engine`_ using Vault's command line client::
+
+  export VAULT_ADDR='http://vaultserver:8200'
+  vault kv put secret/myproject/mybucketkey key=$(dd bs=32 count=1 if=/dev/urandom of=/dev/stdout 2>/dev/null | base64)
+
+Output::
+
+  ====== Metadata ======
+  Key              Value
+  ---              -----
+  created_time     2019-08-29T17:01:09.095824999Z
+  deletion_time    n/a
+  destroyed        false
+  version          1
+
+  === Data ===
+  Key    Value
+  ---    -----
+  key    Ak5dRyLQjwX/wb7vo6Fq1qjsfk1dh2CiSicX+gLAhwk=
+
+The URL to the secret in Vault must be provided in the Gateway configuration
+file (see below).
+
+Configure the Ceph Object Gateway
+=================================
+
+Edit the Ceph configuration file to enable Vault as a KMS for server-side
+encryption::
+
+   rgw crypt s3 kms backend = vault
+   rgw crypt vault auth = token
```

Comment:
You aren't very clear what this "token" is -- but you clearly don't mean the random key you just described creating.  Your test logic uses the root token - but you'd never want to use that in production.  Instead you'd want an ap specific token with "least permissions".  Probably you should have a brief description what the required token permissions are, followed by an example creating a token & policy.

Reply:
So "token" here means we're using the "token-based" authentication method. It implies you need to supply a token in a file, whose path is set in "rgw crypt vault token file". What actual token you have inside that file is up to the user, we only used the root token as an example.

Having said that, we don't think the token-based auth is a good option for real-world deployments but it is the possibly the simplest one and hence the best one for development and testing.

What we plan to actually use in production is the agent-based authentication method, but that we will introduce in the next iteration.